File: python/ftrans/client/minio_client.py
from ftrans.client.base import BaseClient
from minio import Minio


class MinioClient(BaseClient):

    def __init__(self, config, secure=False):
        """
        MinIO Client

        :param config: client config
        :type config: BaseConfig
        :param secure: Flag to indicate to use secure (TLS) connection to S3 service or not. Defaults to False
        :type secure: bool, optional
        """
        self.config = config
        self.secure = secure
        self.client = None
        self.connect()

    def connect(self):
        self.client = Minio('{}:{}'.format(self.config.host_addr,
                                           self.config.host_port),
                            access_key=self.config.access_key,
                            secret_key=self.config.secret_key,
                            secure=self.secure)

    def disconnect(self):
        del self.client
        self.client = None

    def create_dir(self, remote_dir):
        found = self.client.bucket_exists(remote_dir)
        if not found:
            self.client.make_bucket(remote_dir)
        else:
            print("Bucket '{}' already exists".format(remote_dir))

    def list_dir(self, remote_dir):
        objects = self.client.list_objects(bucket_name=remote_dir,)
        return objects

    def upload_file(self,
                    local_filepath,
                    remote_dir,
                    remote_filename,
                    metadata=None):
        result = self.client.fput_object(remote_dir,
                                         remote_filename,
                                         local_filepath,
                                         metadata=metadata)
        return result

    def download_file(self, remote_dir, remote_filename, local_filepath):
        response = self.client.fget_object(remote_dir, remote_filename,
                                           local_filepath)
        return response

    def remove_file(self, remote_dir, remote_filename):
        self.client.remove_object(remote_dir, remote_filename)

    def exists_dir(self, remote_dir):
        return self.client.bucket_exists(remote_dir)

    def exists(self, remote_dir, remote_filename):
        try:
            return self.client.bucket_exists(remote_dir) and \
                self.client.stat_object(remote_dir, remote_filename)
        except Exception as e:
            return False

File: cpp/src/client/minio_client.cc
#include "client/minio_client.h"

#include <fmt/format.h>

namespace ftrans {

MinioClient::MinioClient(const ClientConfig& config, const bool& secure)
    : BaseClient(config), secure_(secure) {
  Connect();
}

MinioClient::~MinioClient() { Disconnect(); }

void MinioClient::Connect() {
  s3_ = std::make_shared<S3Client>(GetEndpoint(), config_.access_key,
                                   config_.secret_key);
}

void MinioClient::Disconnect() { s3_.reset(); }

bool MinioClient::CreateDirectory(const std::string& remote_dir) {
  if (Exists(remote_dir)) {
    return true;
  }
  Minio::S3ClientIO io;
  s3_->MakeBucket(remote_dir, io);
  if (io.Failure()) {
    std::cerr << "ERROR: failed to create bucket" << std::endl;
    std::cerr << "response:\n" << io << std::endl;
    std::cerr << "response body:\n" << io.response.str() << std::endl;
    return false;
  } else {
    return true;
  }
}

bool MinioClient::RemoveDirectory(const std::string& remote_dir) {
  Minio::S3ClientIO io;
  s3_->RemoveBucket(remote_dir, io);
  if (io.Failure()) {
    std::cerr << "ERROR: failed to remove bucket" << std::endl;
    std::cerr << "response:\n" << io << std::endl;
    std::cerr << "response body:\n" << io.response.str() << std::endl;
    return false;
  } else {
    return true;
  }
}

std::vector<std::string> MinioClient::ListDirectory(
    const std::string& remote_dir) {
  Minio::S3ClientIO io;
  s3_->ListObjects(remote_dir, io);
  // TODO: list objects into a vector
  std::cerr << "ERROR: not implemented" << std::endl;
  std::cerr << "response:\n" << io << std::endl;
  std::cerr << "response body:\n" << io.response.str() << std::endl;
  return {};
}

bool MinioClient::UploadFile(const std::string& local_filepath,
                             const std::string& remote_dir,
                             const std::string& remote_filename) {
  Minio::S3ClientIO io;
  s3_->PutObject(remote_dir, remote_filename, local_filepath, io);
  if (io.Failure()) {
    std::cerr << "ERROR: failed to upload file" << std::endl;
    std::cerr << "response:\n" << io << std::endl;
    std::cerr << "response body:\n" << io.response.str() << std::endl;
    return false;
  } else {
    return true;
  }
}

bool MinioClient::DownloadFile(const std::string& remote_dir,
                               const std::string& remote_filename,
                               const std::string& local_filepath) {
  Minio::S3ClientIO io;
  s3_->GetObject(remote_dir, remote_filename, io);
  // TODO: save objects into local file
  std::cerr << "ERROR: not implemented" << std::endl;
  std::cerr << "response:\n" << io << std::endl;
  std::cerr << "response body:\n" << io.response.str() << std::endl;
  return false;
}

bool MinioClient::RemoveFile(const std::string& remote_dir,
                             const std::string& remote_filename) {
  Minio::S3ClientIO io;
  s3_->DeleteObject(remote_dir, remote_filename, io);
  if (io.Failure()) {
    std::cerr << "ERROR: failed to remove file" << std::endl;
    std::cerr << "response:\n" << io << std::endl;
    std::cerr << "response body:\n" << io.response.str() << std::endl;
    return false;
  } else {
    return true;
  }
}

bool MinioClient::Exists(const std::string& remote_dir,
                         const std::string& remote_filename) {
  Minio::S3ClientIO io;
  s3_->ListObjects(remote_dir, io);
  if (io.response.str().find(remote_filename) != std::string::npos) {
    return true;
  } else {
    return false;
  }
}

bool MinioClient::Exists(const std::string& remote_dir) {
  Minio::S3ClientIO io;
  s3_->ListBuckets(io);
  if (io.response.str().find(remote_dir) != std::string::npos) {
    return true;
  } else {
    return false;
  }
}

std::string MinioClient::GetEndpoint() {
  if (secure_) {
    return fmt::format("https://{}:{}", config_.host_addr, config_.host_port);
  } else {
    return fmt::format("http://{}:{}", config_.host_addr, config_.host_port);
  }
}

}  // namespace ftrans